File: src/test/java/cz/vhromada/catalog/service/GameServiceTest.java
package cz.vhromada.catalog.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import cz.vhromada.catalog.domain.Game;
import cz.vhromada.catalog.repository.GameRepository;
import cz.vhromada.catalog.utils.GameUtils;
import cz.vhromada.common.repository.MovableRepository;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.service.MovableServiceTest;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;

/**
 * A class represents test for class {@link GameService}.
 *
 * @author Vladimir Hromada
 */
class GameServiceTest extends MovableServiceTest<Game> {

    /**
     * Instance of {@link GameRepository}
     */
    @Mock
    private GameRepository repository;

    /**
     * Test method for {@link GameService#GameService(GameRepository, Cache)} with null repository for games.
     */
    @Test
    void constructor_NullGameRepository() {
        assertThatThrownBy(() -> new GameService(null, getCache())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link GameService#GameService(GameRepository, Cache)} with null cache.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new GameService(repository, null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableRepository<Game> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Game> getService() {
        return new GameService(repository, getCache());
    }

    @Override
    protected String getCacheKey() {
        return "games";
    }

    @Override
    protected Game getItem1() {
        return GameUtils.newGameDomain(1);
    }

    @Override
    protected Game getItem2() {
        return GameUtils.newGameDomain(2);
    }

    @Override
    protected Game getAddItem() {
        return GameUtils.newGameDomain(null);
    }

    @Override
    protected Game getCopyItem() {
        final Game game = GameUtils.newGameDomain(null);
        game.setPosition(0);

        return game;
    }

    @Override
    protected Class<Game> getItemClass() {
        return Game.class;
    }

    @Override
    protected void assertDataDeepEquals(final Game expected, final Game actual) {
        GameUtils.assertGameDeepEquals(expected, actual);
    }

}

File: src/test/kotlin/cz/vhromada/catalog/service/GameServiceTest.kt
package cz.vhromada.catalog.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.catalog.domain.Game
import cz.vhromada.catalog.repository.GameRepository
import cz.vhromada.catalog.utils.GameUtils
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.service.MovableServiceTest
import cz.vhromada.common.test.utils.TestConstants
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [GameService].
 *
 * @author Vladimir Hromada
 */
class GameServiceTest : MovableServiceTest<Game>() {

    /**
     * Instance of [GameRepository]
     */
    @Mock
    private lateinit var repository: GameRepository

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    override fun getRepository(): JpaRepository<Game, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<Game> {
        return GameService(repository, accountProvider, timeProvider, cache)
    }

    override fun getCacheKey(): String {
        return "games${TestConstants.ACCOUNT.id}"
    }

    override fun getItem1(): Game {
        return GameUtils.newGameDomain(1)
    }

    override fun getItem2(): Game {
        return GameUtils.newGameDomain(2)
    }

    override fun getAddItem(): Game {
        return GameUtils.newGameDomain(null)
    }

    override fun getCopyItem(): Game {
        return GameUtils.newGameDomain(null)
                .copy(position = 0)
    }

    override fun initAllDataMock(data: List<Game>) {
        whenever(repository.findByAuditCreatedUser(any())).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findByAuditCreatedUser(TestConstants.ACCOUNT_ID)
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): Game {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<Game> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: Game, actual: Game) {
        GameUtils.assertGameDeepEquals(expected, actual)
    }

}
